docs: haw-sequencer (#3367)
OffchainLabs/arbitrum-docs: docs/how-arbitrum-works/deep-dives/sequencer.mdx

<ImageZoom src="/img/haw-sequencer-operations.svg" alt="Sequencer operations" className="img-600px" />

## Role of the Sequencer feed in the network

The Sequencer feed serves several vital functions within the Arbitrum ecosystem:

- **State synchronization**: Nodes use the feed to keep their state up to date with the latest network state, ensuring consistency across the decentralized platform.
- **Application development**: Developers can build applications that respond instantly to network events, enabling features such as live updates, instant notifications, and real-time analytics.
- **Ecosystem transparency**: The feed promotes transparency and trust within the community by providing visibility into transaction sequencing and network activity.

## How it works

#### 1. Transaction receipt

When a user submits a transaction via JSON-RPC, it is sent to the `Sequencer.PublishTransaction`, which begins the following process:

1. **Forwarding check**: If the node is not the active sequencer, the transaction is forwarded to the active one. Only the elected sequencer produces blocks.

> 2. **Surplus check**: If `ExpectedSurplusHardThreshold` is set, the L1 pricing surplus must exceed the theshold; otherwise, the transaction is rejected (anti-DoS tied to L1 fee economics).

Reviewer: ```suggestion
2. **Surplus check**: If `ExpectedSurplusHardThreshold` is set, the L1 pricing surplus must exceed the threshold; otherwise, the transaction is rejected (anti-DoS tied to L1 fee economics).
```

3. **Sender allowlist**: If configured, only allowlisted addresses may submit (not active on Arbitrum One, Nova, Sepolia).
4. **Type filtering**: Arbitrum internal transaction types and EIP-4844 blob transactions are rejected.
5. **Timeboost delay**: If Timeboost is active and the current round has an express lane controller, **non-express-lane transactions are delayed** by `ExpressLaneAdvantage` (default: 200ms) before being queued. Express Lane transactions skip this delay.
6. **Queue insertion**: The transaction is wrapped in a `txQueueItem` (containing the transaction, result channel, context, Timeboost metadata) and pushed into the `txQueue` channel (bound by `QueueSize`, default 1024).

#### 2. Block creation loop

The sequencer runs a tight loop (governed by `MaxBlockSpeed`, default 250ms minimal interval) that repeatedly calls `createBlock`:

1. **Drain three queues in priority order**:

- `timeboostAuctionResolutionTxQueue` highest priority (auction resolution transactions)
- `txRetryQueue` transactions retried after nonce/gas issues
- `txQueue` regular incoming transactions

> After the first transaction, it continues draining for upto `ReadFromTxQueueTimeout` (default 10ms) to batch multiple transactions into a single block. 2. **Pre-checks per queued item**:

Reviewer: ```suggestion
After the first transaction, it continues draining for up to `ReadFromTxQueueTimeout` (default 10ms) to batch multiple transactions into a single block.

2. **Pre-checks per queued item**:
```

- Context expiration
- `MaxTxDataSize` (default 95KB)
- `BaseFee` check (reject `GasFeeCap` < block's `BaseFee`)

3. **Nonce pre-checking**:

- Uses an LRU `nonceCache` (address → nonce). Transactions with too-high nonces are placed in the `nonceFailureCache`. When a predecessor later succeeds, waiting transactions are revived and placed in the `txRetryQueue`.

4. **Inactive check**:

- If the sequencer discovers it's no longer active (a forwarder was set), all queued items are forwarded to the new active sequencer.

5. **L1 block/timestamp validation**:

- Ensures the L1 block number is known and the L1 timestamp is within `MaxAcceptableTimestampDelta` (default 1 hour).

6. **Block production**:

- Calls `ExecutionEngine.SequenceTransactions`.

#### 3. Block execution

This is where transactions become a block:

1. Retrieves the current state from the blockchain.
2. Calls `arbos.ProduceBlockAdvanced`, which:

- Iterates through transactions via `hooks.NextTxToSequence`.

> - Applies `preTxFilter` (nonce cache check, conditional options check, address filtering)

Reviewer: ```suggestion
- Applies `preTxFilter` (nonce cache check, conditional options check, address filtering).
```

- **Executes each transaction** against ArbOS state.
- Applies `postTxFilter` (revert gas rejection, nonce cache update, nonce failure revival).
- Enforces per-block data size limits.

3.  Constructs an `L1IncomingMessage` from successfully sequenced transactions.
4.  Generates `blockMetadata`—a bitmask recording which transactions were timeboosted.
5.  Calls `TransactionStreamer.WriteMessageFromSequencer` to persist.
6.  Appends the block to the local blockchain.

#### 4. Persistence & broadcast

1. Verifies this node is still the chosen sequencer.
2. If a `SeqCoordinator` is present, it atomically writes the message to **Redis** and extends the lockout (for high availability replication to standby nodes).
3. Writes the message (with block hash and metadata) to the **local database**.
4. Calls `broadcastMessages` to push the message to the **WebSocket feed** via the `Broadcaster`.

#### 5. L1 posting

The `BatchPoster` runs independently, reading messages from the `TransactionStreamer` in sequence and posting them to L1. The mechanisms that run during this process are:

> - **Compression**: messages are Brotli-compressed with adaptive compression levels based on backlog size.

Reviewer: ```suggestion
- **Compression**: Messages are Brotli-compressed with adaptive compression levels based on backlog size.
```

- **Posting methods**: Calldata (`addSequencerL2BatchFromOrigin`) or EIP-4844 blobs (`addSequencerL2BatchFromBlobs`).
- **DA providers**: Supports AnyTrust committees and blob storage as alternatives to direct L1 posting.
- **L1 bounds tracking**: Ensures batches have valid L1 timestamps.
- **Empty batch management**: Can post empty batches after `MaxEmptyBatchDelay` to advance the chain.

:::note

> This is a high-level summary of how the BatchPoster operates within the overall Sequencing mechanism. To learn more about the BatchPoster refer to the Batch poster deep dive.

Reviewer: ```suggestion
This is a high-level summary of how the BatchPoster operates within the overall Sequencing mechanism. To learn more about the BatchPoster, refer to the [Batch poster deep dive](/how-arbitrum-works/deep-dives/batchposter.mdx).
```

:::

#### 6. Sequencer Coordinator (high-availability)

The `SeqCoordinator` implements Redis-based leader election:

- A `CHOSENSEQ_KEY` in Redis indicates which sequencer URL holds the lock.
- The active sequencer acquires a **time-limited lockout** (default 1 minute) and must periodically refresh it.
- Each sequencer that wants to become active sets a `WantsLockout` key; the highest-priority one is selected.

> - The active sequencer **writes signed messages to Redis** (HMAC-signed). Standby nodes poll Redis and adds messages to their local `TransactionStreamer`.

Reviewer: ```suggestion
- The active sequencer **writes signed messages to Redis** (HMAC-signed). Standby nodes poll Redis and add messages to their local `TransactionStreamer`.
```

- **Activation**: Clears the forwarder, unpauses the sequencer, and syncs the Timeboost state from Redis.
- **Deactivation**: Sets up forwarding to the new active sequencer, pauses local block production.
- **Cleanup**: The active sequencer periodically deletes finalized messages from Redis.

#### 7. Broadcast feed system

The feed distributes sequenced messages to all non-sequencer nodes in real-time:

- The `Broadcaster` runs a WebSocket server, signing each message with a data signer and maintaining a **backlog** for clients connecting mid-stream.
- On startup or lock acquisition, `PopulateFeedBacklog` fills the backlog from the database.
- Receiving nodes call `TransactionStreamer.AddBroadcastMessages`, which validates messages against the local database—feed messages can never reorganize confirmed L1 messages.

<ImageZoom src="/img/haw-sequencer-feed.svg" alt="Sequencer feed" className="img-600px" />

:::note

> It's worth noting that batches posted to the parent chain are the second method of the Broadcast feed system. To learn more refer to the Batch poster deep dive.

Reviewer: ```suggestion
Batches posted to the parent chain are the second method of the Broadcast feed system. To learn more, refer to the Batch poster deep dive.
```

:::

#### 8. Delayed sequencer

Handles messages originating on L1 (deposits, retryable tickets):

> - Monitors the `DelayedBridged` for finalized delayed messages.

Reviewer: ```suggestion
- Monitors the `DelayedBridge` for finalized delayed messages.
```

- Sequences them into the L2 chain via `ExecutionEngine.SequenceDelayedMessage`.

> - Supports **address filtering**—if a delayed messages touches a filtered address, the sequencer halts until the transaction hash is added to an onchain filter.

Reviewer: ```suggestion
- Supports **address filtering**—if a delayed message touches a filtered address, the sequencer halts until the transaction hash is added to an onchain filter.
```

- Only runs on the active sequencer node.

#### 9. Timeboost/Express lane

> Timeboost is a transaction priority auction mechanism. A deep dive can be found on the Timeboost page, but a short summary of how it works:

Reviewer: ```suggestion
Timeboost is a transaction priority auction mechanism. A deep dive can be found on the Timeboost page, but here is a short summary of how it works:
```

- Time is divided into fixed-duration **rounds** (minimum 10 seconds).
- Before each round, an **auction** determines the **express lane controller**—the address whose transactions get priority.
- The express lane controller's transactions **skip the 200ms delay** that normal transactions experience when the express lane is active.

> ##### Flow for Express lane transactions

Reviewer: ```suggestion
##### Flow for Express Lane transactions
```

1. Controller submits via `Sequencer.PublishExpressLaneTransaction`.
2. `ExpressLaneService.ValidateExpressLaneTx` checks sender == controller, correct round, valid chain ID.
3. `ExpressLaneService.SequenceExpressLaneSubmission` inserts into a per-round **ordering queue** (express lane submissions have sequence numbers and are delivered in order).
4. Messages drain in sequence order → `PublishTimeboostedTransaction` → enters `txQueue` with no delay.
5. The `isTimeboosted` flag is preserved through block production into `blockMetadata`.

> ##### Flow for normal transactions when the Express lane is active

Reviewer: ```suggestion
##### Flow for normal transactions when the Express Lane is active
```

1. Transaction arrives at `PublishTransaction`.

> 2. Since `CurrentRoundHasController` returns true, `time.Sleep(ExpressLaneAdvantage)` delay by 200ms.

Reviewer: ```suggestion
2. Since `CurrentRoundHasController` returns true, `time.Sleep(ExpressLaneAdvantage)` delays by 200ms.
```

3. Only after the delay does the transaction enter the queue.

##### Auction system

- The `AuctioneerServer` runs the auction process using EIP-712-signed bids.
- Auctions resolve during the `AuctionClosing` window.
- Resolution transactions get their own highest-priority queue in the sequencer.

## Soft finality and trust model

“Soft finality” refers to the preliminary confirmation of transactions based on the Sequencer’s real-time feed. Key aspects include:

> - **Dependence on Sequencer integrity**: The feed’s accuracy and reliability depend on the Sequencer operating honestly and without significant downtime. > The confirmations are provisional and rely on the Sequencer’s integrity and availability.

Reviewer: ```suggestion
- **Dependence on Sequencer integrity**: The feed's accuracy and reliability depend on the Sequencer operating honestly and without significant downtime.
```

> - **Immediate user feedback**: Users can act on transaction confirmations swiftly, improving the user experience. > Transactions are confirmed almost instantly as they are accepted and ordered by the Sequencer. > Enables applications to offer rapid responses and real-time interactions, enhancing user experience.

Reviewer: ```suggestion
- **Immediate user feedback**: Users can act on transaction confirmations swiftly, improving the user experience.
```

> - **Eventual consistency with the parent chain**: While the real-time feed provides quick updates, ultimate security, and finality are established once transactions are posted to and finalized on the parent chain. Refer to the Finality section for an in-depth discussion.

Reviewer: ```suggestion
- **Eventual consistency with the parent chain**: While the real-time feed provides quick updates, ultimate security and finality are established once transactions are posted to and finalized on the parent chain.
```

- **Application development**: Developers can build applications that respond instantly to network events, enabling features such as live updates, instant notifications, and real-time analytics.
- **Ecosystem transparency**: The feed promotes transparency and trust within the community by providing visibility into transaction sequencing and network activity.

## How it works

#### 1. Transaction receipt

When a user submits a transaction via JSON-RPC, it is sent to the `Sequencer.PublishTransaction`, which begins the following process:

1. **Forwarding check**: If the node is not the active sequencer, the transaction is forwarded to the active one. Only the elected sequencer produces blocks.
2. **Surplus check**: If `ExpectedSurplusHardThreshold` is set, the L1 pricing surplus must exceed the threshold; otherwise, the transaction is rejected (anti-DoS tied to L1 fee economics).
3. **Sender allowlist**: If configured, only allowlisted addresses may submit (not active on Arbitrum One, Nova, Sepolia).
4. **Type filtering**: Arbitrum internal transaction types and EIP-4844 blob transactions are rejected.

> 5. **Timeboost delay**: If Timeboost is active and the current round has an express lane controller, **non-express-lane transactions are delayed** by `ExpressLaneAdvantage` (default: 200ms) before being queued. Express Lane transactions skip this delay.

Reviewer: ```suggestion
5. **Timeboost delay**: If Timeboost is active and the current round has an Express Lane controller, **transactions outside the Express Lane are delayed** by `ExpressLaneAdvantage` (default: 200ms) before being queued. Express Lane transactions skip this delay.
```

6. **Queue insertion**: The transaction is wrapped in a `txQueueItem` (containing the transaction, result channel, context, Timeboost metadata) and pushed into the `txQueue` channel (bound by `QueueSize`, default 1024).

#### 2. Block creation loop

> The sequencer runs a tight loop (governed by `MaxBlockSpeed`, default 250ms minimal interval) that repeatedly calls `createBlock`:

Reviewer: ```suggestion
The Sequencer runs a tight loop (governed by `MaxBlockSpeed`, default 250ms minimal interval) that repeatedly calls `createBlock`:
```

1. **Drain three queues in priority order**:

> - `timeboostAuctionResolutionTxQueue` highest priority (auction resolution transactions)

Reviewer: ```suggestion
- `timeboostAuctionResolutionTxQueue` — highest priority (auction resolution transactions)
```

> - `txRetryQueue` transactions retried after nonce/gas issues

Reviewer: ```suggestion
- `txRetryQueue` — transactions retried after nonce/gas issues
```

> - `txQueue` regular incoming transactions

Reviewer: ```suggestion
- `txQueue` — regular incoming transactions
```

3. **Nonce pre-checking**:

- Uses an LRU `nonceCache` (address → nonce). Transactions with too-high nonces are placed in the `nonceFailureCache`. When a predecessor later succeeds, waiting transactions are revived and placed in the `txRetryQueue`.

4. **Inactive check**:

- If the sequencer discovers it's no longer active (a forwarder was set), all queued items are forwarded to the new active sequencer.

5. **L1 block/timestamp validation**:

- Ensures the L1 block number is known and the L1 timestamp is within `MaxAcceptableTimestampDelta` (default 1 hour).

6. **Block production**:

- Calls `ExecutionEngine.SequenceTransactions`.

#### 3. Block execution

This is where transactions become a block:

1. Retrieves the current state from the blockchain.
2. Calls `arbos.ProduceBlockAdvanced`, which:

- Iterates through transactions via `hooks.NextTxToSequence`.
- Applies `preTxFilter` (nonce cache check, conditional options check, address filtering).
- **Executes each transaction** against ArbOS state.
- Applies `postTxFilter` (revert gas rejection, nonce cache update, nonce failure revival).
- Enforces per-block data size limits.

> 3.  Constructs an `L1IncomingMessage` from successfully sequenced transactions.

Reviewer: ```suggestion
3. Constructs an `L1IncomingMessage` from successfully sequenced transactions.
```

> 4.  Generates `blockMetadata`—a bitmask recording which transactions were timeboosted.

Reviewer: ```suggestion
4. Generates `blockMetadata`—a bitmask recording which transactions were timeboosted.
```

> 5.  Calls `TransactionStreamer.WriteMessageFromSequencer` to persist.

Reviewer: ```suggestion
5. Calls `TransactionStreamer.WriteMessageFromSequencer` to persist.
```

> 6.  Appends the block to the local blockchain.

Reviewer: ```suggestion
6. Appends the block to the local blockchain.
```

3. Writes the message (with block hash and metadata) to the **local database**.
4. Calls `broadcastMessages` to push the message to the **WebSocket feed** via the `Broadcaster`.

#### 5. L1 posting

The `BatchPoster` runs independently, reading messages from the `TransactionStreamer` in sequence and posting them to L1. The mechanisms that run during this process are:

- **Compression**: Messages are Brotli-compressed with adaptive compression levels based on backlog size.
- **Posting methods**: Calldata (`addSequencerL2BatchFromOrigin`) or EIP-4844 blobs (`addSequencerL2BatchFromBlobs`).
- **DA providers**: Supports AnyTrust committees and blob storage as alternatives to direct L1 posting.
- **L1 bounds tracking**: Ensures batches have valid L1 timestamps.
- **Empty batch management**: Can post empty batches after `MaxEmptyBatchDelay` to advance the chain.

:::note

This is a high-level summary of how the BatchPoster operates within the overall Sequencing mechanism. To learn more about the BatchPoster, refer to the [Batch poster deep dive](/how-arbitrum-works/deep-dives/batchposter.mdx).

:::

> #### 6. Sequencer Coordinator (high-availability)

Reviewer: ```suggestion
#### 6. Sequencer Coordinator (high availability)
```

- Each sequencer that wants to become active sets a `WantsLockout` key; the highest-priority one is selected.
- The active sequencer **writes signed messages to Redis** (HMAC-signed). Standby nodes poll Redis and add messages to their local `TransactionStreamer`.
- **Activation**: Clears the forwarder, unpauses the sequencer, and syncs the Timeboost state from Redis.
- **Deactivation**: Sets up forwarding to the new active sequencer, pauses local block production.
- **Cleanup**: The active sequencer periodically deletes finalized messages from Redis.

#### 7. Broadcast feed system

The feed distributes sequenced messages to all non-sequencer nodes in real-time:

- The `Broadcaster` runs a WebSocket server, signing each message with a data signer and maintaining a **backlog** for clients connecting mid-stream.
- On startup or lock acquisition, `PopulateFeedBacklog` fills the backlog from the database.
- Receiving nodes call `TransactionStreamer.AddBroadcastMessages`, which validates messages against the local database—feed messages can never reorganize confirmed L1 messages.

<ImageZoom src="/img/haw-sequencer-feed.svg" alt="Sequencer feed" className="img-600px" />

:::note

> Batches posted to the parent chain are the second method of the Broadcast feed system. To learn more, refer to the Batch poster deep dive.

Reviewer: ```suggestion
Batches posted to the parent chain are the second method of the Broadcast feed system. To learn more, refer to the [Batch poster deep dive](/how-arbitrum-works/deep-dives/batchposter.mdx).
```

:::

#### 8. Delayed sequencer

Handles messages originating on L1 (deposits, retryable tickets):

- Monitors the `DelayedBridge` for finalized delayed messages.
- Sequences them into the L2 chain via `ExecutionEngine.SequenceDelayedMessage`.
- Supports **address filtering**—if a delayed message touches a filtered address, the sequencer halts until the transaction hash is added to an onchain filter.
- Only runs on the active sequencer node.

> #### 9. Timeboost/Express lane

Reviewer: ```suggestion
#### 9. Timeboost/Express Lane
```

> Timeboost is a transaction priority auction mechanism. A deep dive can be found on the Timeboost page, but here is a short summary of how it works:

Reviewer: ```suggestion
Timeboost is a transaction priority auction mechanism. A deep dive can be found on the [Timeboost page](/how-arbitrum-works/timeboost/gentle-introduction.mdx), but here is a short summary of how it works:
```

- Time is divided into fixed-duration **rounds** (minimum 10 seconds).

> - Before each round, an **auction** determines the **express lane controller**—the address whose transactions get priority.

Reviewer: ```suggestion
- Before each round, an **auction** determines the **Express Lane controller**—the address whose transactions get priority.
```

> - The express lane controller's transactions **skip the 200ms delay** that normal transactions experience when the express lane is active.

Reviewer: ```suggestion
- The Express Lane controller's transactions **skip the 200ms delay** that normal transactions experience when the Express Lane is active.
```

##### Flow for Express Lane transactions

1. Controller submits via `Sequencer.PublishExpressLaneTransaction`.

> 2. `ExpressLaneService.ValidateExpressLaneTx` checks sender == controller, correct round, valid chain ID.

Reviewer: ```suggestion
2. `ExpressLaneService.ValidateExpressLaneTx` checks that the sender is the current controller, the round is correct, and the chain ID is valid.
```

> 3. `ExpressLaneService.SequenceExpressLaneSubmission` inserts into a per-round **ordering queue** (express lane submissions have sequence numbers and are delivered in order).

Reviewer: ```suggestion
3. `ExpressLaneService.SequenceExpressLaneSubmission` inserts into a per-round **ordering queue** (Express Lane submissions have sequence numbers and are delivered in order).
```
